SLAM/PoseOptimizationSLAM3D/pose_optimization_slam_3d.py

```diff
@@ -0,0 +1,535 @@
+            q_obj = Quaternion(*q_new)
+            return q_obj
+    
+
+class RotVec:
+
+    def __init__(self, ax=0., ay=0., az=0., quaternion=None):
+        if quaternion is None:
+            self.ax = ax; self.ay = ay; self.az = az
+        else:
+            x = quaternion.qx; y = quaternion.qy; z = quaternion.qz
+            norm_im = np.sqrt(x**2 + y**2 + z**2)
+            if (norm_im < 1e-7):
+                self.ax = 2*x; self.ay = 2*y; self.az = 2*z
+            else:
+                th = 2 * np.arctan2(norm_im, quaternion.qw)
+                th = self.pi2pi(th)
+                self.ax = x / norm_im * th
+                self.ay = y / norm_im * th
+                self.az = z / norm_im * th
+
+    def inverted(self):
+
+        return RotVec(-self.ax, -self.ay, -self.az)
+
+    def toRotationMatrix(self):
+
+        q = self.toQuaternion()
+        q_vec =  np.array([q.qx, q.qy, q.qz]).reshape(3,1)
+        qw = q.qw
+        mat = (qw**2 - q_vec.T @ q_vec) * np.eye(3) + \
+               2 * q_vec @ q_vec.T - 2 * qw * skew_symmetric(q_vec.reshape(-1,))
+        return mat.T
+    
+    def toQuaternion(self):
+
+        v = np.sqrt(self.ax**2 + self.ay**2 + self.az**2)
+        if (v < 1e-6):
+            return Quaternion(1, 0, 0, 0)
+        else:
+            return Quaternion(np.cos(v/2), np.sin(v/2)*self.ax/v,
+                            np.sin(v/2)*self.ay/v, np.sin(v/2)*self.az/v)
+
+    def pi2pi(self, rad):
+
+        val = np.fmod(rad, 2.0 * np.pi)
+        if val > np.pi:
+            val -= 2.0 * np.pi
+        elif val < -np.pi:
+            val += 2.0 * np.pi
+
+        return val
+    
+
+class TripletList:
+
+    def __init__(self):
+        self.row = []
+        self.col = []
+        self.data = []
+
+    def push_back(self, irow, icol, idata):
+        self.row.append(irow)
+        self.col.append(icol)
+        self.data.append(idata)
+
+class Pose3D:
+
+    def __init__(self, x, y, z, qw, qx, qy, qz):
+        self.x = x; self.y = y; self.z = z
+        self.qw = qw; self.qx = qx; self.qy = qy; self.qz = qz
+    
+    def pos(self):
+        v = np.array([self.x, self.y, self.z]).reshape(3,1)
+        return v
+    
+    def rv(self):
+        q = Quaternion(self.qw, self.qx, self.qy, self.qz)
+        return RotVec(quaternion = q)
+
+    def ominus(self, base):
+        t = base.rv().toRotationMatrix().T @ (self.pos() - base.pos())
+        q = base.rv().toQuaternion().conjugate().quat_mult(self.rv().toQuaternion(), out = 'Quaternion')
+        return Pose3D(t[0][0], t[1][0], t[2][0], q.qw, q.qx, q.qy, q.qz)
+
+
+class Constrant3D:
+
+    def __init__(self, id1, id2, t, info_mat):
+        self.id1 = id1
+        self.id2 = id2
+        self.t = t
+        self.info_mat = info_mat
+
+def plot_nodes(nodes, ax, color ="-r", label = ""):
+
+    x, y, z = [], [], []
+    for n in nodes:
+        x.append(n.x); y.append(n.y); z.append(n.z)
+    ax.plot(x, y, z, color, label=label)
```

Comment:
@rsasaki0109  If you put this code here:
```
    x = np.array(x)
    y = np.array(y)
    z = np.array(z)
    max_range = np.array([x.max()-x.min(), y.max()-y.min(), z.max()-z.min()]).max() / 2.0
    mid_x = (x.max()+x.min()) * 0.5
    mid_y = (y.max()+y.min()) * 0.5
    mid_z = (z.max()+z.min()) * 0.5
    ax.set_xlim(mid_x - max_range, mid_x + max_range)
    ax.set_ylim(mid_y - max_range, mid_y + max_range)
    ax.set_zlim(mid_z - max_range, mid_z + max_range)

```

you can get the correctly scaled 3D graph
![SS 2019-09-07 at 14 38 16](https://user-images.githubusercontent.com/3813847/64470392-665d0480-d17d-11e9-9b96-ae2d3d02bc3f.png)

Reply:
Sorry. Because my PC is frozen and I tried it with another PC, It took some time.